docs/Config-File.md
This page contains the configuration structure for HuskSync.

## Configuration structure
📁 `plugins/HuskSync/`
- 📄 `config.yml`: General plugin configuration
- 📄 `server.yml`: Server ID configuration
- 📄 `messages-xx-xx.yml`: Plugin locales, formatted in MiniMessage (see [[Translations]])

## Example files
<details>
<summary>config.yml</summary>

```yaml
# ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
# ┃        HuskSync Config       ┃
# ┃    Developed by William278   ┃
# ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
# ┣╸ Information: https://william278.net/project/husksync
# ┣╸ Config Help: https://william278.net/docs/husksync/config-file/
# ┗╸ Documentation: https://william278.net/docs/husksync

# Locale of the default language file to use.
# Docs: https://william278.net/docs/husksync/translations
language: en-gb
# Whether to automatically check for plugin updates on startup
check_for_updates: true
# Specify a common ID for grouping servers running HuskSync. Don't modify this unless you know what you're doing!
cluster_id: ''
# Enable development debug logging
debug_logging: true
# Whether to enable the Player Analytics hook.
# Docs: https://william278.net/docs/husksync/plan-hook
enable_plan_hook: true
# Whether to cancel game event packets directly when handling locked players if ProtocolLib or PacketEvents is installed
cancel_packets: true
# Add HuskSync commands to this list to prevent them from being registered (e.g. ['userdata'])
disabled_commands: []
# Database settings
database:
  # Type of database to use (MYSQL, MARIADB, POSTGRES, MONGO)
  type: MYSQL
  # Specify credentials here for your MYSQL, MARIADB, POSTGRES OR MONGO database
  credentials:
    host: localhost
    port: 3306
    database: minecraft
    username: root
    password: ''
    # Only change this if you're using MARIADB or POSTGRES
    parameters: ?autoReconnect=true&useSSL=false&useUnicode=true&characterEncoding=UTF-8
  # MYSQL, MARIADB, POSTGRES database Hikari connection pool properties. Don't modify this unless you know what you're doing!
  connection_pool:
    maximum_pool_size: 10
    minimum_idle: 10
    maximum_lifetime: 1800000
    keepalive_time: 0
    connection_timeout: 5000
  # Advanced MongoDB settings. Don't modify unless you know what you're doing!
  mongo_settings:
    using_atlas: false
    parameters: ?retryWrites=true&w=majority&authSource=HuskSync
  # Names of tables to use on your database. Don't modify this unless you know what you're doing!
  table_names:
    users: husksync_users
    user_data: husksync_user_data
# Redis settings
redis:
  # Specify the credentials of your Redis server here.
  # Set "user" to '' if you don't have one or would like to use the default user.
  # Set "password" to '' if you don't have one.
  credentials:
    host: localhost
    port: 6379
    # Only change the database if you know what you are doing. The default is 0.
    database: 0
    user: ''
    password: ''
    use_ssl: false
  # Options for if you're using Redis sentinel. Don't modify this unless you know what you're doing!
  sentinel:
    # The master set name for the Redis sentinel.
    master: ''
    # List of host:port pairs
    nodes: []
    password: ''
# Data syncing settings
synchronization:
  # The data synchronization mode to use (LOCKSTEP or DELAY). LOCKSTEP is recommended for most networks.
  # Docs: https://william278.net/docs/husksync/sync-modes
  mode: LOCKSTEP
  # The number of data snapshot backups that should be kept at once per user
  max_user_data_snapshots: 16
  # Number of hours between new snapshots being saved as backups (Use "0" to backup all snapshots)
  snapshot_backup_frequency: 4
  # List of save cause IDs for which a snapshot will be automatically pinned (so it won't be rotated).
  # Docs: https://william278.net/docs/husksync/data-rotation#save-causes
  auto_pinned_save_causes:
    - INVENTORY_COMMAND
    - ENDERCHEST_COMMAND
    - BACKUP_RESTORE
    - LEGACY_MIGRATION
    - MPDB_MIGRATION
  # Whether to create a snapshot for users on a world when the server saves that world
  save_on_world_save: true
  # Configuration for how and when to sync player data when they die
  save_on_death:
    # Whether to create a snapshot for users when they die (containing their death drops)
    enabled: true
    # What items to save in death snapshots? (DROPS or ITEMS_TO_KEEP). Note that ITEMS_TO_KEEP (suggested for keepInventory servers) requires a Paper 1.19.4+ server.
    items_to_save: ITEMS_TO_KEEP
    # Should a death snapshot still be created even if the items to save on the player's death are empty?
    save_empty_items: true
    # Whether dead players who log out and log in to a different server should have their items saved.
    sync_dead_players_changing_server: true
  # Whether to use the snappy data compression algorithm. Keep on unless you know what you're doing
  compress_data: true
  # Where to display sync notifications (ACTION_BAR, CHAT or NONE)
  notification_display_slot: ACTION_BAR
  # Persist maps locked in a Cartography Table to let them be viewed on any server
  persist_locked_maps: true
  # If using the DELAY sync method, how long should this server listen for Redis key data updates before pulling data from the database instead (i.e., if the user did not change servers).
  network_latency_milliseconds: 500
  # Which data types to synchronize.
  # Docs: https://william278.net/docs/husksync/sync-features
  features:
    inventory: true
    ender_chest: true
    experience: true
    advancements: true
    game_mode: true
    flight_status: true
    potion_effects: true
    statistics: true
    health: true
    hunger: true
    attributes: true
    persistent_data: true
    location: false
  # Commands which should be blocked before a player has finished syncing (Use * to block all commands)
  blacklisted_commands_while_locked:
    - '*'
  # Configuration for how to sync attributes
  attributes:
    # Which attribute types should be saved as part of attribute syncing. Supports wildcard matching.
    # (e.g. ['minecraft:generic.max_health', 'minecraft:generic.*'])
    synced_attributes: 
      - "minecraft:generic.max_health"
      - "minecraft:max_health"
      - "minecraft:generic.max_absorption"
      - "minecraft:max_absorption"
      - "minecraft:generic.luck"
      - "minecraft:luck"
      - "minecraft:generic.scale"
      - "minecraft:scale"
      - "minecraft:generic.step_height"
      - "minecraft:step_height"
      - "minecraft:generic.gravity"
      - "minecraft:gravity"
    # Which attribute modifiers should not be saved when syncing users. Supports wildcard matching.
    # (e.g. ['minecraft:effect.speed', 'minecraft:effect.*'])
    ignored_modifiers: ['minecraft:effect.*', 'minecraft:creative_mode_*']
  # Event priorities for listeners (HIGHEST, NORMAL, LOWEST, MONITOR). Change if you encounter plugin conflicts. MONITOR is not recommended, only use it for plugins that need HuskSync to run after them.
  event_priorities:
    quit_listener: LOWEST
    join_listener: LOWEST
    death_listener: NORMAL
```

</details>

<details>
<summary>server.yml</summary>

```yaml
# ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
# ┃     HuskSync - Server ID     ┃
# ┃    Developed by William278   ┃
# ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
# ┣╸ This file should contain the ID of this server as defined in your proxy config.
# ┗╸ If you join it using /server alpha, then set it to 'alpha' (case-sensitive)

name: beta
```

</details>

common/src/main/java/net/william278/husksync/config/Settings.java
/*
 * This file is part of HuskSync, licensed under the Apache License 2.0.
 *
 *  Copyright (c) William278 <will27528@gmail.com>
 *  Copyright (c) contributors
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */

package net.william278.husksync.config;

import com.google.common.collect.Lists;
import de.exlll.configlib.Comment;
import de.exlll.configlib.Configuration;
import lombok.AccessLevel;
import lombok.Getter;
import lombok.NoArgsConstructor;
import net.william278.husksync.command.PluginCommand;
import net.william278.husksync.data.DataSnapshot;
import net.william278.husksync.data.Identifier;
import net.william278.husksync.database.Database;
import net.william278.husksync.listener.EventListener;
import net.william278.husksync.sync.DataSyncer;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.List;
import java.util.Locale;
import java.util.Map;

/**
 * Plugin settings, read from config.yml
 */
@SuppressWarnings("FieldMayBeFinal")
@Getter
@Configuration
@NoArgsConstructor(access = AccessLevel.PRIVATE)
public class Settings {

    protected static final String CONFIG_HEADER = """
            ┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
            ┃        HuskSync Config       ┃
            ┃    Developed by William278   ┃
            ┣━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
            ┣╸ Information: https://william278.net/project/husksync
            ┣╸ Config Help: https://william278.net/docs/husksync/config-file/
            ┗╸ Documentation: https://william278.net/docs/husksync""";

    // Top-level settings
    @Comment({"Locale of the default language file to use.", "Docs: https://william278.net/docs/husksync/translations"})
    private String language = Locales.DEFAULT_LOCALE;

    @Comment("Whether to automatically check for plugin updates on startup")
    private boolean checkForUpdates = true;

    @Comment("Specify a common ID for grouping servers running HuskSync. "
             + "Don't modify this unless you know what you're doing!")
    private String clusterId = "";

    @Comment("Enable development debug logging")
    private boolean debugLogging = false;

    @Comment({"Whether to enable the Player Analytics hook.", "Docs: https://william278.net/docs/husksync/plan-hook"})
    private boolean enablePlanHook = true;

    @Comment("Whether to cancel game event packets directly when handling locked players if ProtocolLib or PacketEvents is installed")
    private boolean cancelPackets = true;

    @Comment("Add HuskSync commands to this list to prevent them from being registered (e.g. ['userdata'])")
    @Getter(AccessLevel.NONE)
    private List<String> disabledCommands = Lists.newArrayList();

    // Database settings
    @Comment("Database settings")
    private DatabaseSettings database = new DatabaseSettings();

    @Getter
    @Configuration
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class DatabaseSettings {

        @Comment("Type of database to use (MYSQL, MARIADB, POSTGRES, MONGO)")
        private Database.Type type = Database.Type.MYSQL;

        @Comment("Specify credentials here for your MYSQL, MARIADB, POSTGRES OR MONGO database")
        private DatabaseCredentials credentials = new DatabaseCredentials();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class DatabaseCredentials {
            private String host = "localhost";
            private int port = 3306;
            private String database = "HuskSync";
            private String username = "root";
            private String password = "pa55w0rd";
            @Comment("Only change this if you're using MARIADB or POSTGRES")
            private String parameters = String.join("&",
                    "?autoReconnect=true", "useSSL=false",
                    "useUnicode=true", "characterEncoding=UTF-8");
        }

        @Comment("MYSQL, MARIADB, POSTGRES database Hikari connection pool properties. Don't modify this unless you know what you're doing!")
        private PoolSettings connectionPool = new PoolSettings();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class PoolSettings {
            private int maximumPoolSize = 10;
            private int minimumIdle = 10;
            private long maximumLifetime = 1800000;
            private long keepaliveTime = 0;
            private long connectionTimeout = 5000;
        }

        @Comment("Advanced MongoDB settings. Don't modify unless you know what you're doing!")
        private MongoSettings mongoSettings = new MongoSettings();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class MongoSettings {
            private boolean usingAtlas = false;
            private String parameters = String.join("&",
                    "?retryWrites=true", "w=majority",
                    "authSource=HuskSync");
        }

        @Comment("Names of tables to use on your database. Don't modify this unless you know what you're doing!")
        @Getter(AccessLevel.NONE)
        private Map<String, String> tableNames = Database.TableName.getDefaults();

        @Comment("Whether to run the creation SQL on the database when the server starts. Don't modify this unless you know what you're doing!")
        private boolean createTables = true;

        @NotNull
        public String getTableName(@NotNull Database.TableName tableName) {
            return tableNames.getOrDefault(tableName.name().toLowerCase(Locale.ENGLISH), tableName.getDefaultName());
        }
    }

    // Redis settings
    @Comment("Redis settings")
    private RedisSettings redis = new RedisSettings();

    @Getter
    @Configuration
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class RedisSettings {

        @Comment({"Specify the credentials of your Redis server here.",
                "Set \"user\" to '' if you don't have one or would like to use the default user.",
                "Set \"password\" to '' if you don't have one."})
        private RedisCredentials credentials = new RedisCredentials();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class RedisCredentials {
            private String host = "localhost";
            private int port = 6379;
            @Comment("Only change the database if you know what you are doing. The default is 0.")
            private int database = 0;
            private String user = "";
            private String password = "";

            @Comment("Use SSL/TLS for encrypted connections.")
            private boolean useSsl = false;

            @Comment("Connection timeout in milliseconds.")
            private int connectionTimeout = 2000;

            @Comment("Socket (read/write) timeout in milliseconds.")
            private int socketTimeout = 2000;

            @Comment("Max number of connections in the pool.")
            private int maxTotalConnections = 50;

            @Comment("Max number of idle connections in the pool.")
            private int maxIdleConnections = 8;

            @Comment("Min number of idle connections in the pool.")
            private int minIdleConnections = 2;

            @Comment("Enable health checks when borrowing connections from the pool.")
            private boolean testOnBorrow = true;

            @Comment("Enable health checks when returning connections to the pool.")
            private boolean testOnReturn = true;

            @Comment("Enable periodic idle connection health checks.")
            private boolean testWhileIdle = true;

            @Comment("Min evictable idle time (ms) before a connection is eligible for eviction.")
            private long minEvictableIdleTimeMillis = 60000;

            @Comment("Time (ms) between eviction runs.")
            private long timeBetweenEvictionRunsMillis = 30000;

            @Comment("Number of retries for commands when connection fails.")
            private int maxRetries = 3;

            @Comment("Base backoff time in ms for retries (exponential backoff multiplier).")
            private int retryBackoffMillis = 200;
        }

        @Comment("Options for if you're using Redis sentinel. Don't modify this unless you know what you're doing!")
        private RedisSentinel sentinel = new RedisSentinel();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class RedisSentinel {
            @Comment("The master set name for the Redis sentinel.")
            private String master = "";
            @Comment("List of host:port pairs")
            private List<String> nodes = Lists.newArrayList();
            private String password = "";
        }

    }

    // Synchronization settings
    @Comment("Data syncing settings")
    private SynchronizationSettings synchronization = new SynchronizationSettings();

    @Getter
    @Configuration
    @NoArgsConstructor(access = AccessLevel.PRIVATE)
    public static class SynchronizationSettings {

        @Comment({"The data synchronization mode to use (LOCKSTEP or DELAY). LOCKSTEP is recommended for most networks.",
                "Docs: https://william278.net/docs/husksync/sync-modes"})
        private DataSyncer.Mode mode = DataSyncer.Mode.LOCKSTEP;

        @Comment("The number of data snapshot backups that should be kept at once per user")
        private int maxUserDataSnapshots = 16;

        @Comment("Number of hours between new snapshots being saved as backups (Use \"0\" to backup all snapshots)")
        private int snapshotBackupFrequency = 4;

        @Comment({"List of save cause IDs for which a snapshot will be automatically pinned (so it won't be rotated).",
                "Docs: https://william278.net/docs/husksync/data-rotation#save-causes"})
        @Getter(AccessLevel.NONE)
        private List<String> autoPinnedSaveCauses = List.of(
                DataSnapshot.SaveCause.INVENTORY_COMMAND.name(),
                DataSnapshot.SaveCause.ENDERCHEST_COMMAND.name(),
                DataSnapshot.SaveCause.BACKUP_RESTORE.name(),
                DataSnapshot.SaveCause.LEGACY_MIGRATION.name(),
                DataSnapshot.SaveCause.MPDB_MIGRATION.name()
        );

        @Comment("Whether to create a snapshot for users on a world when the server saves that world")
        private boolean saveOnWorldSave = true;

        @Comment("Configuration for how and when to sync player data when they die")
        private SaveOnDeathSettings saveOnDeath = new SaveOnDeathSettings();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class SaveOnDeathSettings {
            @Comment("Whether to create a snapshot for users when they die (containing their death drops)")
            private boolean enabled = false;

            @Comment("What items to save in death snapshots? (DROPS or ITEMS_TO_KEEP). "
                     + "Note that ITEMS_TO_KEEP (suggested for keepInventory servers) requires a Paper 1.19.4+ server.")
            private DeathItemsMode itemsToSave = DeathItemsMode.DROPS;

            @Comment("Should a death snapshot still be created even if the items to save on the player's death are empty?")
            private boolean saveEmptyItems = true;

            @Comment("Whether dead players who log out and log in to a different server should have their items saved.")
            private boolean syncDeadPlayersChangingServer = true;

            /**
             * Represents the mode of saving items on death
             */
            public enum DeathItemsMode {
                DROPS,
                ITEMS_TO_KEEP
            }
        }

        @Comment("Whether to use the snappy data compression algorithm. Keep on unless you know what you're doing")
        private boolean compressData = true;

        @Comment("Where to display sync notifications (ACTION_BAR, CHAT or NONE)")
        private Locales.NotificationSlot notificationDisplaySlot = Locales.NotificationSlot.ACTION_BAR;

        @Comment("Persist maps locked in a Cartography Table to let them be viewed on any server")
        private boolean persistLockedMaps = true;

        @Comment("How long to wait with the DELAY sync method before checking Redis for data. Not used by the LOCKSTEP sync method.")
        private int networkLatencyMilliseconds = 500;

        @Comment({"How long to wait for player saves to finish during server shutdown before closing the database connections",
                "Should be set to a value between 5000-50000ms, to not consume the server watchdog timeout (typically 60000ms)."})
        private int shutdownSaveTimeoutMilliseconds = 5000;

        @Comment({"Which data types to synchronize.", "Docs: https://william278.net/docs/husksync/sync-features"})
        @Getter(AccessLevel.NONE)
        private Map<String, Boolean> features = Identifier.getConfigMap();

        @Comment("Commands which should be blocked before a player has finished syncing (Use * to block all commands)")
        private List<String> blacklistedCommandsWhileLocked = new ArrayList<>(List.of("*"));

        @Comment("Prevent ender pearl launch events from being blocked while the user is being locked (fixes stasis chambers on 1.21.1+)")
        private boolean allowPearlSpawningWhileLocked = true;

        @Comment("Configuration for how to sync attributes")
        private AttributeSettings attributes = new AttributeSettings();

        @Getter
        @Configuration
        @NoArgsConstructor(access = AccessLevel.PRIVATE)
        public static class AttributeSettings {

            @Comment({"Which attribute types should be saved as part of attribute syncing. Supports wildcard matching.",
                    "(e.g. ['minecraft:generic.max_health', 'minecraft:generic.*'])"})
            @Getter(AccessLevel.NONE)
            private List<String> syncedAttributes = new ArrayList<>(List.of(
                    "minecraft:generic.max_health", "minecraft:max_health",
                    "minecraft:generic.max_absorption", "minecraft:max_absorption",
                    "minecraft:generic.luck", "minecraft:luck",
                    "minecraft:generic.scale", "minecraft:scale",
                    "minecraft:generic.step_height", "minecraft:step_height",
                    "minecraft:generic.gravity", "minecraft:gravity"
            ));

            @Comment({"Which attribute modifiers should be saved. Supports wildcard matching.",
                    "(e.g. ['minecraft:effect.speed', 'minecraft:effect.*'])"})
            @Getter(AccessLevel.NONE)
            private List<String> ignoredModifiers = new ArrayList<>(List.of(
                    "minecraft:effect.*", "minecraft:creative_mode_*"
            ));

            private boolean matchesWildcard(@NotNull String pat, @NotNull String value) {
                if (!pat.contains(":")) {
                    pat = "minecraft:%s".formatted(pat);
                }
                if (!value.contains(":")) {
                    value = "minecraft:%s".formatted(value);
                }
                return pat.contains("*") ? value.matches(pat.replace("*", ".*")) : pat.equals(value);
            }

            public boolean isIgnoredAttribute(@NotNull String attribute) {
                return syncedAttributes.stream().noneMatch(wildcard -> matchesWildcard(wildcard, attribute));
            }

            public boolean isIgnoredModifier(@NotNull String modifier) {
                return ignoredModifiers.stream().anyMatch(wildcard -> matchesWildcard(wildcard, modifier));
            }

        }

        @Comment({"Event priorities for listeners (HIGHEST, NORMAL, LOWEST, MONITOR). Change if you encounter plugin conflicts",
                "MONITOR is not recommended, only use it for plugins that need HuskSync to run after them, e.g. combat loggers"})
        @Getter(AccessLevel.NONE)
        private Map<String, String> eventPriorities = EventListener.ListenerType.getDefaults();

        @Comment("Enable check-in petitions for data syncing (don't change this unless you know what you're doing)")
        private boolean checkinPetitions = false;

        public boolean doAutoPin(@NotNull DataSnapshot.SaveCause cause) {
            return autoPinnedSaveCauses.contains(cause.name());
        }

        public boolean isFeatureEnabled(@NotNull Identifier id) {
            return id.isCustom() || features.getOrDefault(id.getKeyValue(), id.isEnabledByDefault());
        }

        @NotNull
        public EventListener.Priority getEventPriority(@NotNull EventListener.ListenerType type) {
            try {
                return EventListener.Priority.valueOf(eventPriorities.get(type.name().toLowerCase(Locale.ENGLISH)));
            } catch (IllegalArgumentException e) {
                return EventListener.Priority.NORMAL;
            }
        }
    }

    public boolean isCommandDisabled(@NotNull PluginCommand command) {
        return disabledCommands.stream().map(c -> c.startsWith("/") ? c.substring(1) : c)
                .anyMatch(c -> c.equalsIgnoreCase(command.getName()) || command.getAliases().contains(c));
    }


}
